internal/pkg/agent/cmd/enroll_cmd.go

```diff
@@ -395,10 +395,6 @@ func (c *enrollCmd) prepareFleetTLS() error {
 	}
 	if c.options.FleetServer.Cert == "" && c.options.FleetServer.CertKey == "" {
 		if c.options.FleetServer.Insecure {
-			// running insecure, force the binding to localhost (unless specified)
```

Comment:
What was the original reason this was done? Just covering the case where the host was unset? Is there a test we can add to keep this bug from happening again?

Reply:
I don't know why this was done here. I'll try to add a couple unit tests for this method.

I think we need to do some minor cleanup around bootstapping (in another pr) as we have both this method and `createFleetServerBootstrapConfig` altering config at different locations

Reply:
cc @blakerouse

Reply:
This was added for security reasons, we didn't want the Fleet Server to be exposed outside of localhost insecure mode.

We only allow it to be exposed in insecure mode when the user is specific and provide a host. I think we should really think about this before we make this change.

Reply:
Reading the issue description it sounds like when it was explicitly configured to bind to `0.0.0.0` it binds to localhost anyway.

Does `c.options.FleetServer.Host` evaluate to `""` when the host was set to `0.0.0.0` intentionally?

Reply:
I think the issue is that the default `0.0.0.0` is not explicitly passed when starting a fleet-server instance.
The `createFleetServerBootstrapConfig` may insert the value if it's blank, but that method is called after  `prepareFleetTLS`.

I think the issue here is that the behaviour of the intended security block is not working as people expect for 8.6+ (even though it is now doing what was intended)

Reply:
So this fixes it to do the correct thing? If you bootstrap with insecure it will only bind to localhost by default?

Reply:
no, it keeps the current (insecure) behaviour.
The issue is that the fleet server is binding to localhost when it's being upgraded from a previous version so remote agents are no longer able to report in

Reply:
So confused because this change was added in 7.13 and its been there for a while? Can you show how the code has changed in 8.6 for this to cause a problem? Sorry I am really lost, and don't understand how this fixes the issue without relaxing security.

Reply:
I'm not sure what exactly caused the issue; in 8.5.3 if we install an agent with the `--fleet-server-insecure-http` flag (and no`--url`), it will bind the external port to `0.0.0.0:8220`
On 8.6.0 it will bind to `localhost:8220`.

Reply:
To me then 8.5.3 was broken and 8.6 is not broken.

But this is also not change in the code, and doesn't look to be an issue in this code path - https://github.com/elastic/elastic-agent/blob/8.5/internal/pkg/agent/cmd/enroll_cmd.go#L396

8.5 code looks the same

Reply:
If we want to keep this behaviour we should highlight that some policy settings are ignored, for instance the fleet-server integration may be:
```yml
- type: fleet-server
  server:
    port: 8220
    host: 0.0.0.0
```
Do we need to add documentation to state that the `server.host` from the policy is ignored if the server is enrolled with `--fleet-server-insecure-http` and that a user must use `--fleet-server-host` if they want any value other then `localhost`?

We should also test that upgrade behaviour for non-default values for the `host`/`port` values are respected (as a follow up item)

Reply:
I think the intent here from both the code and Blake's comments is to bind exclusively to localhost _only_ when the host is not specified. Specifying `host: 0.0.0.0` should not force binding to localhost like it does now.

If someone intentionally wants a fleet server with insecure HTTP to bind to 0.0.0.0 or some other interface/host they should be able to do that.

Forcing localhost when there is no host specified is correct from a security perspective so that we don't end up with fleet servers with insecure configurations exposed on public IPs accidentally.

It seems like we really haven't found the true cause of this bug yet since this exact same code path worked properly prior to 8.6.0. The correct behaviour existed in 8.5.3 but we seem to have broken it for unexplained reasons.

Reply:
If you look at the report in the linked issue (https://github.com/elastic/elastic-agent/issues/2197) they are claiming that they explicitly configure the host to `0.0.0.0` and it now ignores it and binds to `localhost` with this behaviour starting in 8.6.0. We perhaps should test this and confirm given the confusion about what it is supposed to do.

There are completely valid reasons to configure the server like this and it should work, for example fleet-server could be behind a load balancer or proxy that does TLS termination.

Reply:
@cmacknz to clarify, I ran into this issue by setting the host in the Fleet Server **integration**, through Kibana:

```yaml
# kibana.yml
# ...
        inputs:
          - type: fleet-server
            vars:
              - name: host
                value: 0.0.0.0
```

If I configure the host [directly in the Elastic Agent](https://github.com/deviantony/docker-elk/commit/48340a904682fc687a9cda936dcb33b049d16235), it works.

I think Kibana's value is ignored because Elastic Agent has no way to tell whether the user explicitly requested `0.0.0.0` or simply relied on the default, which happens to be `0.0.0.0` as well.

Reply:
The Kibana value is always ignored for Fleet Server, and it has always been that way. Only the values used during the bootstrap process is what is used for Fleet Server.

Reply:
@blakerouse in that case I'm fine with having to explicitly set the host _in the Elastic Agent_, since the root cause of the issue is my assumption that Kibana's values were actually taken into account.

It's still confusing to upgrade from 8.5 and 8.6 and realize that things aren't working anymore, but this could have easily been addressed with a changelog entry.

Reply:
Should we consider removing the ability to alter the host value in Kibana to make it clear that it's ignored?

Reply:
IMO yes, since this isn't obvious at all.